baseplate/diagnostics/tracing/publisher.py

```diff
@@ -0,0 +1,191 @@
+import argparse
+import logging
+import time
+
+import requests
+
+from . import MAX_SPAN_SIZE, MAX_QUEUE_SIZE
+from baseplate import config, metrics_client_from_config
+from baseplate._compat import configparser
+from baseplate.events.publisher import BatchFull, V1Batch, TimeLimitedBatch
+from baseplate.message_queue import MessageQueue, TimedOutError
+
+
+logger = logging.getLogger(__name__)
+
+# seconds to wait for response from Zipkin server
+POST_TIMEOUT_DEFAULT = 3
+
+# maximum size (in bytes) of a batch of traces
+MAX_BATCH_SIZE_DEFAULT = 500 * 1024
+
+# seconds to wait between retries
+RETRY_DELAY_DEFAULT = 2
+# maximum number of retries when publishing traces
+RETRY_LIMIT_DEFAULT = 10
+
+ZIPKIN_API_VERSION_DEFAULT = "v1"
+ZIPKIN_SCHEME_DEFAULT = "http"
+
+
+class RetryLimitExceeded(Exception):
+    pass
+
+
+class TraceBatch(V1Batch):
+    def __init__(self, max_size):
+        super(TraceBatch, self).__init__(max_size=max_size)
+
+
+class ZipkinPublisher(object):
+    """Zipkin trace publisher."""
+    def __init__(self, endpoint,
+            metrics_client,
+            post_timeout=POST_TIMEOUT_DEFAULT,
+            zipkin_api_version=ZIPKIN_API_VERSION_DEFAULT,
+            zipkin_scheme=ZIPKIN_SCHEME_DEFAULT,
+            retry_delay=RETRY_DELAY_DEFAULT,
+            retry_limit=RETRY_LIMIT_DEFAULT,
+            num_conns=5):
+
+        adapter = requests.adapters.HTTPAdapter(pool_connections=num_conns,
+                                                pool_maxsize=num_conns)
+        self.session = requests.Session()
+        self.session.mount("{}://".format(zipkin_scheme), adapter)
+        self.endpoint = "{}://{}/api/{}/spans".format(zipkin_scheme, endpoint, zipkin_api_version)
+        self.metrics = metrics_client
+        self.post_timeout = post_timeout
+        self.retry_delay = retry_delay
+        self.retry_limit = retry_limit
+
+    def publish(self, payload):
+        """Publish spans to Zipkin API.
+
+        :param baseplate.events.publisher.SerializedBatch payload: Count and
+            payload to publish.
+        """
+        if not payload.count:
+            return
+
+        logger.info("Sending batch of %s traces", payload.count)
+        headers = {
+            "User-Agent": "baseplate-trace-publisher/1.0",
+            "Content-Type": "application/json",
+        }
```

Comment:
Is there anything in Zipkin to allow us to tell it which service we are to aid in ratelimiting / diagnosis of bad actors?

Reply:
Spans are tied to the service using the [``service_name`` config](https://github.com/reddit/baseplate/blob/master/baseplate/__init__.py#L80). I don't see a way to the publishing of a set of spans to the publisher.

Reply:
no prob, thanks!